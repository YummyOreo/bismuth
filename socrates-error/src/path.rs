#![allow(unused)]
use std::{io::stdout, path::PathBuf};

use crossterm::{
    cursor,
    event::{read, Event, KeyCode},
    execute,
    style::{Color, Print, ResetColor, SetForegroundColor},
    terminal, Result,
};

use crate::error_ui;

fn update_input(path: &mut String) -> Option<()> {
    if let Event::Key(k) = read().expect("Should be able to read") {
        match k.code {
            KeyCode::Char(x) => {
                path.push(x);
                execute!(stdout(), Print(x.to_string()));
            }
            KeyCode::Backspace => {
                path.pop();
                let print_path = format!("Enter new path: {}", path);
                execute!(
                    stdout(),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::MoveToColumn(0),
                    Print(print_path),
                );
            }
            KeyCode::Enter => return Some(()),
            _ => {}
        }
    }
    None
}

pub fn md_file_error(_description: &str) -> Option<PathBuf> {
    let options = vec![
        // "Make folder at path.".to_string(),
        "Chose another path".to_string(),
    ];

    let option = error_ui(&options)?;

    match option {
        // 1 => {}
        1 => {
            execute!(
                stdout(),
                cursor::MoveUp(u16::try_from(options.len()).unwrap() + 1_u16),
                terminal::Clear(terminal::ClearType::CurrentLine),
                cursor::SavePosition
            );
            for i in 0..options.len() + 1 {
                execute!(
                    stdout(),
                    cursor::MoveDown(1),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::RestorePosition,
                );
            }

            execute!(
                stdout(),
                ResetColor,
                Print("Enter new path: "),
                cursor::Show
            );
            let mut path = "".to_string();
            loop {
                if update_input(&mut path).is_some() {
                    break;
                }
            }
            return Some(PathBuf::from(path));
        }
        _ => panic!("Something has gone terribly wrong"),
    }
    None
}
