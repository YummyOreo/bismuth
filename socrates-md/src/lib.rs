#![allow(dead_code)]
use std::fs;
use std::path::PathBuf;

pub mod load;

#[derive(Debug)]
pub struct MarkdownFile {
    pub content: String,
    path: PathBuf,
}

#[derive(Debug)]
pub enum MarkdownFileError {
    IsFileError(String),
    NotMarkdownError(String),
    ErrorLoadingFile(String),

    NotDirectoryError(String),
}

impl MarkdownFile {
    pub fn load_file(path: &PathBuf) -> Result<Self, MarkdownFileError> {
        if !path.is_file() {
            return Err(MarkdownFileError::IsFileError(
                path.to_string_lossy().to_string(),
            ));
        }

        if let Some(s) = path.extension() {
            let s = s.to_ascii_lowercase();
            let s = s.to_str().expect("Should be string");

            if matches!(s, "md" | "markdown") {
                return Ok(MarkdownFile {
                    path: path.to_path_buf(),
                    content: fs::read_to_string(path).expect("file should be there"),
                });
            }
        }
        return Err(MarkdownFileError::NotMarkdownError(
            path.to_string_lossy().to_string(),
        ));
    }
}

// # Tests:

#[cfg(test)]
mod test {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_load() {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path("./testdata/output/");
        let snapshot =
            MarkdownFile::load_file(&PathBuf::from("./testdata/tests/test.markdown")).unwrap();
        settings.bind(|| {
            insta::assert_snapshot!(format!("{:?}", snapshot));
        });
    }

    #[test]
    fn test_load_1() {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path("./testdata/output/");
        let snapshot =
            MarkdownFile::load_file(&PathBuf::from("./testdata/tests/test1.MARKDOWN")).unwrap();
        settings.bind(|| {
            insta::assert_snapshot!(format!("{:?}", snapshot));
        });
    }

    #[test]
    fn test_load_2() {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path("./testdata/output/");
        let snapshot =
            MarkdownFile::load_file(&PathBuf::from("./testdata/tests/test3.md")).unwrap();
        settings.bind(|| {
            insta::assert_snapshot!(format!("{:?}", snapshot));
        });
    }

    #[test]
    fn test_load_3() {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_path("./testdata/output/");
        let snapshot =
            MarkdownFile::load_file(&PathBuf::from("./testdata/tests/TEST4.MD")).unwrap();
        settings.bind(|| {
            insta::assert_snapshot!(format!("{:?}", snapshot));
        });
    }
}
