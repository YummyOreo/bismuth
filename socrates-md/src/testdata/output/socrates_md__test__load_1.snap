---
source: socrates-md/src/lib.rs
expression: "format!(\"{:?}\", snapshot)"
---
MarkdownFile { content: "# this is test 1\r\n", path: "./testdata/tests/test1.MARKDOWN" }
